design/019-local_development_flow.md

```diff
@@ -0,0 +1,55 @@
+- Keep working, providing in ``conan.tools.xxxx`` new tools and toolchains a full working set of helpers to use in the ``generate()`` method, that operate creating files for the user convenience, so users can build directly with their native tools, instead of relying on the ``conan build` command.
+- One of two scenarios could happen: If the above works always, then the ``conan build`` command might become completely useless and will be removed. If on the contrary, there are still some use cases that cannot fit the ``generate()`` approach, Conan 2.0 will make the ``conan build`` complete, not using any saved state representation from a previous install. The 3 files ``conaninfo.txt``, ``conanbuildinfo.txt`` and ``graph_info.json`` will not be generated at all locally. The ``conan build`` command will accept the same arguments as ``conan install``, settings, options, profiles, lockfiles and will compute the graph of dependencies and provide the exact same state to the ``build()`` method as if the package was being created in the cache.
+
+NOTE: There are a couple of side use cases, like ``source`` and ``imports`` functionality, that can be relevant but are not as core as the one presented here. Some other proposals can happen regarding these other elements, let's leave those out of this proposal and discussion.
+
+
+## Detailed Design
+
+In the case that the ``conan build`` command is maintained, its only minor inconvenience could be that it would be slower than the current version, as it will evaluate the whole graph of dependencies. This shouldn't be a blocker for some reasons:
+
+- For small and medium size projects, this is not relevant at all. For large projects, with hundreds of packages in the dependency graph, it might take a few seconds.
+- The ``conan build`` command will not be necessary in most cases. There is a lot of effort ongoing on the new toolchains approach to integrate build systems to have native builds, that do not require Conan to build locally, and thus can use the IDE and other native tools, providing a better developer experience.
+- The time used to compute the whole graph, for large graphs, is still in the order of seconds, which can be typically small compared with actual build times.
+- Conan 2.0 will implement a new graph model. There are some ideas to improve the graph resolution performance, it is possible that the time to evaluate dependency graphs might be reduced.
+
+
+``conan export-pkg`` will be simplified, as it will always execute the ``package()`` method. For existing external binaries, it could be just a ``self.copy(“*”)``, if the artifacts are already arranged in the final layout. But as the local “packaged” state is no longer a use case, the functionality of ``export-pkg`` that was skipping the ``package()`` method is no longer necessary.
+
+
+## Migration notes
+These commands are not frequently used in automated or CI flows, mostly developer side commands. No migration issues expected.
```

Comment:
Every CI workflow at our establishment is using `conan build`. Not every CI flow in an a Conan application produces a package, remember there are projects that can be pure consumers.

Reply:
Good point, ``conan build`` might be maintained if there is enough support, not a problem. Just that it will operate slightly differently regarding the saved state and the arguments. I would like to understand, if the ``generate()`` abstraction is good enough, wouldn't it be interesting to have a "native build" calling CMake or Visual or whatever directly instead of ``conan build``?

Reply:
> Just that it will operate slightly differently regarding the saved state and the arguments.

Could I still give it an install folder to get the state from?

> wouldn't it be interesting to have a "native build" calling CMake or Visual or whatever directly instead of conan build?

In some cases, but then it changes from

`invoke build`

to

* If you're on this project then use `scons`
* And if you're on this project, you can use `cmake --build`
* But if you want to be more direct, then you need to figure out if the underlying CMake generator was `Unix Makefiles` or `Xcode` or `Visual Studio`
* ...

I mean, we have some people who'll reach for an IDE and know they can find the project in the build folder, and some people who just hit the text files with an editor, and then know they can `invoke build`.

Reply:
We are also using `conan build` in our CI when testing that packages can be built correctly on PRs, without polluting the CI server's cache. I find it much more elegant than using native tools as it makes CI script much simpler and leaves the complexity of using special build tool for each specific platform. We abstract away the builds for iOS/Android/Windows/... in conan package that all our packages `python_requires`. This gives us also the ability to have versioning on build procedure (i.e. old versions supported only ios device builds, new also support Mac Catalyst builds). By having that in our CI script we would be at risk of using incompatible build procedure (i.e. new Mac Catalyst build) on older version of the package, yielding incomprehensible error.

Reply:
Good, I see the case, and how it can be convenient.

> Could I still give it an install folder to get the state from?

This is exactly what we are trying to get rid of. That state is really problematic, always buggy and incomplete, it is really a nightmare. The proposal is to pass ``conan build`` the arguments that you will pass to have that state, like you do to ``conan install``, mostly the ``--profile``. Just use the same arguments as ``conan install``, or skip ``conan install``, as it can be already callled by ``conan build``, as it will expand the graph anyway to recover the state.

The performance penalty should be very reduced, our benchmark for large graphs with hundreds of deps (when already installed in the cache) is a few seconds, which is typically much lower than the build itself, and negligible in CI time.

Reply:
I like the idea of skipping `conan install` and just using `conan build` :+1: